### PR TITLE
fix/routing: simplify promotion/demotion and reduce number

### PR DESCRIPTION
**Pull request by jeanphilippeD**
We checked for promotion demotion everytime a churn event occured.
If we have some in the backlog, we risked changing multiple times in
quick sucession.

The new mechanism has multiple benefit:
 - Only update elders once if multiple node in churn backlog.
 - Clear interaction between the different polling (events, relocation,
   backloged churn, and Elder churn).
 - Avoid duplication that could break shared state in adult.
 - Shorten the flow since promote Demote does not happen as a result
   of an other event completing.

Test:
clippy + soak test

### CODE REVIEW

File: src/chain/chain.rs [hunk 1]
```diff
@@ -540,15 +548,19 @@ impl Chain {
...
     /// Returns a set of EldersInfos to vote for.
-    pub fn promote_and_demote_elders(&mut self) -> Result<Vec<EldersInfo>, RoutingError> {
-        self.assert_no_prefix_change("promote and demote");
+    fn promote_and_demote_elders(&mut self) -> Result<Option<Vec<EldersInfo>>, RoutingError> {
+        if !self.members_changed || !self.can_poll_churn() {
+            // Nothing changed that could impact elder set, or we cannot process it yet.
+            return Ok(None);
+        }
 
         // TODO: the split decision should be based on the number of all members, not just elders.
```
Comment (madadam) [inline src/chain/chain.rs]: Not part of this PR, but this comment is outdated and can be deleted.

File: src/chain/chain.rs [hunk 2]
```diff
@@ -540,15 +548,19 @@ impl Chain {
 
     /// Generate a new section info based on the current set of members.
     /// Returns a set of EldersInfos to vote for.
-    pub fn promote_and_demote_elders(&mut self) -> Result<Vec<EldersInfo>, RoutingError> {
-        self.assert_no_prefix_change("promote and demote");
+    fn promote_and_demote_elders(&mut self) -> Result<Option<Vec<EldersInfo>>, RoutingError> {
+        if !self.members_changed || !self.can_poll_churn() {
+            // Nothing changed that could impact elder set, or we cannot process it yet.
+            return Ok(None);
```
Comment (madadam) [inline src/chain/chain.rs]: Why not just returning an empty vec?

Comment (madadam) [reply src/chain/chain.rs]: Or maybe change the return type to something like:
```rust
enum NewEldersInfos {
    None,
    One(EldersInfo),
    Two(EldersInfo),
}
```
(but with better names)

Comment (jeanphilippeD) [reply src/chain/chain.rs]: The empty vec make the call site/intent less clear (poll_accumulated), so in the end I converted it to an option.

The One, Tow would also make the call site (futher up) less clear as instead of looping through all the given info we'd have to have specific code for either case.

File: src/chain/chain.rs [hunk 3]
```diff
@@ -66,6 +66,8 @@ pub struct Chain {
     churn_in_progress: bool,
     /// Marker indicating we are processing a relocation.
     relocation_in_progress: bool,
+    /// Marker indicating that elders may need to change,
```
Comment (maqi) [inline src/chain/chain.rs]: can the comment be more explicit, say `Marker indicating elders may need to change, as there is membership change`

File: src/chain/chain.rs (hunk 2 above)
Comment (madadam) [reply src/chain/chain.rs]: You can `impl IntoIterator` for the `NewEldersinfos` thing if you want to be really fancy :) But I'm OK with it as is for now. We can always give it a refactoring pass some time later if we want.

File: src/chain/chain.rs (hunk 3 above)
Comment (jeanphilippeD) [reply src/chain/chain.rs]: It may be because of anything, that would change elders, in particular, if a member become Relocating. so I'm not sure it make sense to be more specific? (It just happen that both membership change and age increase happen on the same events)

File: src/chain/chain.rs [hunk 4]
```diff
@@ -354,7 +357,7 @@ impl Chain {
             _ => false,
         };
 
-        if start_churn_event && (self.churn_in_progress || self.relocation_in_progress) {
+        if start_churn_event && !self.can_poll_churn() {
```
Comment (maqi) [inline src/chain/chain.rs]: just to confirm that the `can_poll_churn` function will check two additional flag `handled_genesis_event` and `split_in_progress` from state.
Is this ok?

File: src/chain/chain.rs [hunk 5]
```diff
@@ -433,11 +436,7 @@ impl Chain {
     fn poll_relocation(&mut self) -> Option<RelocateDetails> {
         // Delay relocation until all backlogged churn events have been handled and no
         // additional churn is in progress. Only allow one relocation at a time.
-        if self.relocation_in_progress
-            || self.churn_in_progress
-            || !self.state.churn_event_backlog.is_empty()
-            || !self.state.handled_genesis_event
-        {
+        if !self.can_poll_churn() || !self.state.churn_event_backlog.is_empty() {
```
Comment (maqi) [inline src/chain/chain.rs]: just to confirm that the `can_poll_churn` function will check additional flag `split_in_progress` from state.
Is this ok?

File: src/chain/chain.rs (hunk 4 above)
Comment (jeanphilippeD) [reply src/chain/chain.rs]: Yes, it should be the same condition that allow an Online/Offline event to be coming out of that function

File: src/chain/chain.rs (hunk 5 above)
Comment (jeanphilippeD) [reply src/chain/chain.rs]: Yes, really split_in_progress will only be true if churn_in_progress is true also, so does not make much difference.
This set of flags can be cleaned up / improved later.
